### PR TITLE
[MRG] API: Delete scan from BIDS dataset

### PR DESCRIPTION
**Pull request by adam2392**
PR Description
--------------

Closes scans.tsv portion of: #439 

~~This is not done yet, but I've setup a unit test to illustrate when this would be used, and also made the preliminary function to stick into the utils API. Please let me know if this looks good to proceed and I can do the same for `def update_participants_tsv()`.~~

Edit: This now implements a `delete_scan` function for the mne_bids API, which safely deletes a scan from both the `scans.tsv` file and the bids_root. It also has informative error messages. The unit test was added to test error message and proper deletion.

This should properly conform to the bids-validator PR: https://github.com/bids-standard/bids-validator/pull/976.

I believe there is currently still an error related to #459 which needs to be solved before CIs are green.

Merge checklist
---------------

Maintainer, please confirm the following before merging:

- [ ] All comments resolved
- [ ] This is not your own PR
- [ ] All CIs are happy
- [ ] PR title starts with [MRG]
- [ ] [whats_new.rst](https://github.com/mne-tools/mne-bids/blob/master/doc/whats_new.rst) is updated
- [ ] PR description includes phrase "closes <#issue-number>"
- [ ] Commit history does not contain any merge commits

### CODE REVIEW

File: mne_bids/utils.py [hunk 1]
```diff
@@ -26,7 +26,67 @@
...
 
-from mne_bids.tsv_handler import _to_tsv, _tsv_to_str
+from mne_bids.tsv_handler import _to_tsv, _tsv_to_str, _from_tsv
+
+
+def update_scans_tsv(bids_root, subject=None, session=None, verbose=True):
+    """Update the scans tsv files inside bids root.
+
+    If there are extra rows in the tsv file
+    that are not present in the BIDs dataset, then it will
```
Comment (agramfort) [inline mne_bids/utils.py:36]:
```suggestion
    that are not present in the BIDS dataset, then it will
```

File: doc/api.rst [hunk 2]
```diff
@@ -41,6 +41,7 @@ Utils (:py:mod:`mne_bids.utils`)
    print_dir_tree
    get_entity_vals
    get_kinds
+   update_scans_tsv
```
Comment (agramfort) [inline doc/api.rst:44]:
```suggestion
   update_scans
```

update_scans is clear enough to me.

**Comment by jasmainak** [comment]
update seems to be a bit of a misnomer to me, don't you think? Our write function is actually doing updates.

what you really want is something to delete an entry from `scans.tsv` that will also remove the files, right?

**Comment by adam2392** [comment]
> update seems to be a bit of a misnomer to me, don't you think? Our write function is actually doing updates.
> 
> what you really want is something to delete an entry from `scans.tsv` that will also remove the files, right?

So this function wouldn’t be doing that and would just be keeping the scans and database up to date. A “sync”. 

What you described sounds like the “delete” in CRUD terms. Would a delete_scan be desired?

**Comment by jasmainak** [comment]
I'm thinking that the user should never manually remove files because BIDS is a complex data structure. They should use the `mne-bids` to remove items.

Also if the contents of `scans.tsv` does not match the actual file structure, does the validator complain? If not, we should raise an issue

**Comment by adam2392** [comment]
> I'm thinking that the user should never manually remove files because BIDS is a complex data structure. They should use the `mne-bids` to remove items. So the right workflow is not manually remove + sync but mne_bids.remove(bids_basename)

That's fair. So perhaps the function(s) to implement should be:

1. `def remove_scan(bids_basename, bids_root)` and

2. `def remove_participant(participant_id, bids_root)`?

**Comment by adam2392** [comment]
> Also if the contents of `scans.tsv` does not match the actual file structure, does the validator complain? If not, we should raise an issue

Yep, I ran it just now w/ the unit test inside here and the validator did not throw an error.

**Comment by agramfort** [comment]
>  the validator did not throw an error.

I think it should...

before implementing many public functions can list what we plan
to need? Just to see

>

**Comment by adam2392** [comment]
> I think it should... before implementing many public functions can list what we plan to need? Just to see

For me, it would be of interest to be able to maintain a list of say... subjects or scans that I don't want in my dataset updated. I can then ideally remove those subjects, or scans from the bids dataset. You suggested we need basic CRUD for certain files.

- C: From an API perspective, `write_raw_bids` is like "adding a scan" into the dataset (and also a participant potentially). 
- R: We have `read_raw_bids` 
- U: I think this is probably the hardest, since allowing arbitrary updating is not easy. 
- D: @jasmainak is suggesting to do this programatically with `remove_scan` and `remove_participant` functions. 

I  think the purpose of this PR is to enable the "D" part of CRUD, removing particular scans, or participants. This could be encapsulated into one function. Idk what to name it, perhaps `delete_bids_files` with initial support for subject and scans?

**Comment by agramfort** [comment]
what do you expect to be able to delete? a full participant? a scan only? a
run only?
would passing a pattern to glob with * be reasonable to list files to
delete?

**Comment by jasmainak** [comment]
I think globbing would be tough. You need the user to pass reasonable patterns not sidecars and stuff. That's why I suggested bids_basename but it's true it doesn't cover use cases such as deleting entire subjects.

**Comment by adam2392** [comment]
> what do you expect to be able to delete? a full participant? a scan only? a run only? would passing a pattern to glob with * be reasonable to list files to delete?

I would say two use cases mainly:

1. A full participant, updating the participants file.
2. A full scan, updating the scans.tsv file and getting rid of related sidecar files

I think a pattern w/ glob could work...:

1. `delete_files(pattern=subject)`
2. `delete_files(pattern=scan)`

The only issue is how do you update the correct files accordingly? Deleting a subject should result in participants.tsv being updated.  Deleting a scan should result in the scans.tsv file. But with a general glob, you might be able to just say... delete the electrodes.tsv file.

**Comment by jasmainak** [comment]
I'm sorry I dropped the ball on this one. Maybe @hoechenberger or @sappelhoff want to take a look too?

File: mne_bids/write.py [hunk 3]
```diff
@@ -1085,11 +1085,10 @@ def write_raw_bids(raw, bids_basename, bids_root, events_data=None,
     unit = UNITS.get(ext, 'n/a')
     manufacturer = MANUFACTURERS.get(ext, 'n/a')
 
-    # save all meta data
+    # save all participants meta data
     _participants_tsv(raw, subject_id, participants_tsv_fname, overwrite,
                       verbose)
     _participants_json(participants_json_fname, True, verbose)
-    _scans_tsv(raw, op.join(kind, bids_fname), scans_fname, overwrite, verbose)
```
Comment (adam2392) [inline mne_bids/write.py:1092]: moved to the bottom of the whole function in case, conversion takes place.

File: mne_bids/write.py [hunk 4]
```diff
@@ -1120,15 +1119,16 @@ def write_raw_bids(raw, bids_basename, bids_root, events_data=None,
...
         raise FileExistsError('"%s" already exists. Please set '  # noqa: F821
                               'overwrite to True.' % bids_fname)
-    _mkdir_p(os.path.dirname(bids_fname))
+    _mkdir_p(os.path.dirname(op.join(data_path, bids_fname)))
 
+    # set the raw file name to now be the absolute path to ensure the files
+    # are placed in the right location
+    bids_fname.prefix = data_path
+
+    # whether or not to convert to RECOMMENDED file formats
```
Comment (adam2392) [inline mne_bids/write.py:1131]: Swapped the order with `_mkdir_p` and the test worked.

**Comment by jasmainak** [comment]
@adam2392 can you make the tests pass?

File: doc/api.rst [hunk 5]
```diff
@@ -56,6 +56,7 @@ Utils
    get_entity_vals
    get_kinds
    BIDSPath
+   delete_scan
```
Comment (jasmainak) [inline doc/api.rst:59]: I don't think this should be in utils, it should be above somewhere in the same place you have write and read functions

File: mne_bids/utils.py [hunk 6]
```diff
@@ -28,7 +28,72 @@
...
+from mne_bids.tsv_handler import (_to_tsv, _tsv_to_str,
+                                  _from_tsv, _drop)
+
+
+def delete_scan(bids_basename, bids_root, verbose=True):
+    """Safely delete a scan inside bids root.
+
+    Deleting a scan that conforms to the bids-validator, will
+    delete both the row in the `scans.tsv` file, and also
+    the corresponding sidecar files and the data file itself.
```
Comment (jasmainak) [inline mne_bids/utils.py:40]: what about `participants.tsv`?

**Comment by adam2392** [comment]
> @adam2392 can you make the tests pass?

@jasmainak Sorry, trying, but there's a larger issue here that I can't get a handle on rn :(. See: https://github.com/bids-standard/bids-validator/issues/1002

You can try it out locally and you'll see the problem w/ CTF and BTI (directory based data files).

File: mne_bids/utils.py (hunk 6 above)
Comment (adam2392) [reply mne_bids/utils.py:40]: Deleting scan shouldn't delete participants.tsv entry unless it's the absolute last scan for that participant. With that being said tho... that should be incorporated. 

Do we want `delete_participant` as part of the `write` API as well then?

File: mne_bids/utils.py [hunk 7]
```diff
@@ -28,7 +28,7 @@
 from mne.time_frequency import psd_array_welch
 
 from mne_bids.config import BIDS_PATH_ENTITIES, reader
-from mne_bids.tsv_handler import _to_tsv, _tsv_to_str
+from mne_bids.tsv_handler import (_to_tsv, _tsv_to_str)
```
Comment (jasmainak) [inline mne_bids/utils.py:31]:
```suggestion
from mne_bids.tsv_handler import _to_tsv, _tsv_to_str
```

File: mne_bids/write.py [hunk 8]
```diff
@@ -1372,3 +1382,109 @@ def write_anat(bids_root, subject, t1w, session=None, acquisition=None,
...
+        delete_participant(bids_basename.subject,
+                           bids_root=bids_root,
+                           verbose=verbose)
+    else:
+        # resave the scans.tsv file
+        scans_tsv = _drop(scans_tsv, bids_fname, 'filename')
+        _to_tsv(scans_tsv, scans_fpath)
+
+
+def delete_participant(subject, bids_root, verbose=True):
```
Comment (jasmainak) [inline mne_bids/write.py:1458]: I'm a bit uncertain of an API like this. In the future will you also need a `delete_session` function?

Couldn't we instead have some checks? E.g., remove all the files that match a certain regular expression/wildcard/keyword argument pair (subject='blah', run='blah' etc.) ? Then finally sanitize the directories + scans.tsv + participants.tsv.

Comment (hoechenberger) [reply mne_bids/write.py:1458]: This would be possible now thanks to `BIDSPath.match()`.

**Comment by adam2392** [comment]
Okay so based on the big changes going on in recent PRs, should this still be useful? Say:

```Python
def delete_matching_entities(root, subject=None, session=None, ...):
# creates a BIDSPath object and gets matching basenames

# even possibly prompt the user if they want to delete with kwarg ``prompt=False``?

# if files are deleted, update scans.tsv

# if subjects are deleted, update participants.tsv
```

**Comment by adam2392** [comment]
> BIDSPath(root=, subject='01').rm() ? deletes all data from subject 1? BIDSPath(root=, datatype='meg').rm() ? deletes all meg data?

Fair! Closing this then.

**Comment by jasmainak** [comment]
> BIDSPath(root=, subject='01').rm() ?

> deletes all data from subject 1?

there is still the issue that you need to update `participants.tsv` and `scans.tsv` ...

**Comment by adam2392** [comment]
> > BIDSPath(root=, subject='01').rm() ?
> 
> > deletes all data from subject 1?
> 
> there is still the issue that you need to update `participants.tsv` and `scans.tsv` ...

Ah... right you are... Okay should "delete" be a function of `BIDSPath` or a separate function?

**Comment by agramfort** [comment]
yes you need to keep the files consistent

>

**Comment by sappelhoff** [comment]
I liked the proposal from https://github.com/mne-tools/mne-bids/issues/458 a lot

**Pull request closed without merging**